### ISSUE TITLE
build failing on CentOS 6.6, gcc 4.4.7

### ISSUE THREAD

**Issue by sebastianelsner** [OpenImageIO/oiio#1145]
Since -Werror is on by default, I get this failiure:

```
cc1plus: warnings being treated as errors
In function ‘void test_special()’:
cc1plus: error: dereferencing pointer ‘values.200’ does break strict-aliasing rules
/workspace/OpenImageIO/oiio/src/include/OpenImageIO/simd.h:688: note: initialized from here
cc1plus: error: dereferencing pointer ‘values.202’ does break strict-aliasing rules
/workspace/OpenImageIO/oiio/src/include/OpenImageIO/simd.h:702: note: initialized from here
make[2]: *** [src/libutil/CMakeFiles/simd_test.dir/simd_test.cpp.o] Error 1
make[1]: *** [src/libutil/CMakeFiles/simd_test.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by hobbes1069** [comment]
I tried reproducing the error but I'm guessing you have a newer boost installed since 1.41 in RHEL/EPEL isn't new enough (1.42 or newer required).

**Comment by sebastianelsner** [comment]
Yes, I am using Boost 1.55.

**Comment by hobbes1069** [comment]
Ok, I was able to build 1.57 from the Fedora Rawhide source RPM with a little tweaking. OIIO 1.5.14 built fine for me. Are you building 1.5.15 or master?

**Comment by sebastianelsner** [comment]
I am building from master with "sse4.2" enabled. 1.5.14 fails, too.
Did you build with the rawhide srpm on CentOS?

**Comment by hobbes1069** [comment]
Yes, I'm the Fedora packager for OIIO so I tend to work from source RPMS, it's required when using mock to build for other platforms anyway. I run Fedora 21 at home but can build for Fedora 20->Rawhide and EPEL 6 & 7 using mock which creates a chroot with all the packages and dependencies for the target system. The only part of the host system that makes it into the chroot is the kernel, which VERY rarely has an effect on the build.

I'll have to try adding the sse option and try again, by default I think only SSE2 is assumed on distro builds as that's all that is guaranteed on x86_64.

**Comment by lgritz** [comment]
I don't see why boost would factor into this.

The lines it cites in simd.h only come into play when built with USE_SIMD=sse4.2, not the default x86_64 sse2.

It doesn't seem to like the pointer cast on those lines. It's definitely what I want to do, but it may be a strange thing that the compiler doesn't like.

I think the best thing is to figure out how to make lines 788 and 702 (and maybe 716 and 730) not to trigger that error.

Since I can't seem to repro, you'll need to help. Please try changing line 702

        simd_t a = _mm_castpd_si128 (_mm_load_sd ((const double *)values));

to

        const char *c = (const char *)values;
        simd_t a = _mm_castpd_si128 (_mm_load_sd ((const double *)c));

In other words, cast it through a char*. You'll need to do something similar to the other lines I mentioned above.

If that works, let me know and I'll submit a fix.

**Comment by hobbes1069** [comment]
No, I wasn't indicating boost was an issue, it's just have the version in RHEL is too old so I had to build it anyway.

Ok, updated my local package to 1.5.15 and enabled sse4.2. My build completed so it may be down to some configuration problem? I have both clang and C11 options disabled.

**Comment by sebastianelsner** [comment]
@lgritz I changed the lines, but it just moved the warning one line down.

```
cc1plus: warning: dereferencing pointer ‘c.202’ does break strict-aliasing rules
/workspace/OpenImageIO/oiio/src/include/OpenImageIO/simd.h:703: note: initialized from here
```

@hobbes1069 If you build for EPEL does it use gcc 4.4? I have googled a bit and found some users complainging that this is a bug in gcc: http://stackoverflow.com/questions/2114378/trying-to-understand-a-gcc-error

If it really is a compiler bug, I would just turn off the -Werror and not care.

**Comment by hobbes1069** [comment]
Yes, everything is built with packages from RHEL and EPEL, only the kernel remains Fedora.

I saw the same thing but I can't reproduce the error. Would my build log help?

**Comment by lgritz** [comment]
Reviewing old tickets...

Has this been resolved, or is it still an issue?

**Comment by sebastianelsner** [comment]
the problem still exists with gcc4.4 + sse4.2. I stayed with the standard sse2, while I hope to move to a more current compiler soon in our facility. You can close it.

**Comment by lgritz** [comment]
OK, thanks.
